docs/event_listeners.md

```diff
@@ -0,0 +1,235 @@
+use async_trait::async_trait;
+
+/// Define a simple event listener that ticks every MS milliseconds.
+pub struct Ticker<const MS: usize> {
+    additional_delay: u64,
+}
+```
+
+Then, define a context for the event listener:
+
+```rust
+#[derive(Copy, Clone)]
+pub struct MyContext {
+    pub additional_delay: u64,
+}
+```
+
+Next, implement `EventListener` for `Ticker`:
+
+```rust
+/// Implement the [`EventListener`] trait for the Ticker struct. [`EventListener`] has two type parameters:
+/// 
+/// - T: is the type of the event that the listener listens for. This can be anything that is Send + Sync + 'static.
+/// In this case, we are using [`Instant`], which is a timestamp.
+///
+/// - Ctx: is the context type that the listener receives when constructed. This can be anything that is Send + Sync + 'static,
+/// with the special requirement that it is the *first* listed additional parameter in the [`job`] macro (i.e., a parameter not
+/// in params(...)). In this case, we are using [`MyContext`].
+#[async_trait]
+impl<const MS: usize> EventListener<Instant, MyContext> for Ticker<MS> {
+    /// Use a reference to the context, [`MyContext`], to construct the listener
+    async fn new(context: &MyContext) -> Result<Self, Error>
+    where
+        Self: Sized,
+    {
+        Ok(Self { additional_delay: context.additional_delay })
+    }
+
+    /// Implement the logic that looks for events. In this case, we have a simple stream
+    /// that returns after MS milliseconds.
+    async fn next_event(&mut self) -> Option<Instant> {
+        tokio::time::sleep(tokio::time::Duration::from_millis(MS as u64 + self.additional_delay)).await;
+        Some(Instant::now())
+    }
+
+    /// After next_event is called, the event gets passed here. This is where you would implement
+    /// listener-specific logic.
+    async fn handle_event(&mut self, _event: Instant) -> Result<(), Error> {
+        Ok(())
+    }
+}
+```
+
+Finally, register the event listener inside the `job` macro using `event_listener`:
+
+```rust
+#[job(
+    id = 0,
+    params(x),
+    result(_),
+    event_listener(Ticker::<6000>), // <-- Register the event listener here
+    verifier(evm = "IncredibleSquaringBlueprint")
+)]
+pub fn hello_event_listener(
+    x: u64,
+    context: MyContext, // <-- The context type must be the first additional parameter
+    env: GadgetConfiguration<parking_lot::RawRwLock>,
+) -> Result<u64, Infallible> {
+    Ok(x.saturating_pow(2u32))
+}
+```
+
+## Wrapper Event Listeners
+Various wrappers exist to simplify how common operations are performed. For example, there is the `PeriodicEventListener`:
+
+### PeriodicEventListener
+Some programs may only be interested in checking for events at regular intervals. In this case, the `PeriodicEventListener` can be used to simplify the process.
+
+A `PeriodicEventListener` is a wrapper that takes 4 type parameters:
+
+* `MSEC`: The number of milliseconds between each event check.
```

Comment:
nit: The `Ticker` example above used `MS`, this one uses `MSEC`

Reply:
Adjusted the example above to use `MSEC` so it's all consistent.